tiledb/sm/rest/rest_client.cc

```diff
@@ -528,6 +528,155 @@ Status RestClient::submit_query_to_rest(const URI& uri, Query* query) {
   return Status::Ok();
 }
 
+bool RestClient::check_tile_alignment(Query* query) {
+  // Only applicable for remote global order writes
+  if (query->type() != QueryType::WRITE ||
+      query->layout() != Layout::GLOBAL_ORDER) {
+    return true;
+  }
+  uint64_t cell_num_per_tile =
+      query->array_schema().dense() ?
+          query->array_schema().domain().cell_num_per_tile() :
+          query->array_schema().capacity();
+
+  if (query_remote_buffer_storage_.empty()) {
+    for (const auto& name : query->buffer_names()) {
+      bool is_var = query->array_schema().var_size(name);
+      bool is_nullable = query->array_schema().is_nullable(name);
+      uint64_t data_size = datatype_size(query->array_schema().type(name));
+      uint64_t cell_size = is_var ? constants::cell_var_offset_size : data_size;
+
+      query_remote_buffer_storage_[name].cell_size = cell_size;
+      // Construct and move a preallocated Buffer to initialize cache.
+      query_remote_buffer_storage_[name].buffer = {
+          cell_num_per_tile * cell_size};
+      if (is_nullable) {
+        query_remote_buffer_storage_[name].buffer_validity = {
+            cell_num_per_tile};
+      }
+    }
+  }
+
+  bool submit = true;
```

Comment:
I think I hit all your points here... I did change `cache_read`->`cache_write` since this is for writes, but if this was not a typo I can rename. Thanks for the review! :slightly_smiling_face: